scripts/hail_batch/eqtl_hail_batch/generate_eqtl_spearman.py

```diff
@@ -0,0 +1,195 @@
+"""Run Spearman rank correlation on SNPs and expression residuals"""
+
+import os
+import hail as hl
+import hailtop.batch as hb
+import pandas as pd
+import numpy as np
+import statsmodels.api as sm
+import statsmodels.stats.multitest as multi
+from patsy import dmatrices  # pylint: disable=no-name-in-module
+from scipy.stats import spearmanr
+
+INPUT_BUCKET = 'cpg-tob-wgs-test'
+OUTPUT_BUCKET = 'cpg-tob-wgs-test'
+DRIVER_IMAGE = os.getenv(
+    'DRIVER_IMAGE',
+    'australia-southeast1-docker.pkg.dev/analysis-runner/images/driver:d2a9c316d6d752edb27623542c8a062db4466842-hail-0.2.73.devc6f6f09cec08',  # noqa: E501; pylint: disable=line-too-long
+)
+
+
+def get_number_of_scatters():
+    """get index of total number of genes"""
+    expression_df = pd.read_csv(
+        f'gs://{INPUT_BUCKET}/kat/input/Plasma_expression.tsv', sep='\t'
+    )
+    # Remove genes with 0 expression in all samples
+    expression_df = expression_df.loc[:, (expression_df != 0).any(axis=0)]
+    # Number of individuals with non-zero expression
+    genes_not_equal_zero = expression_df.iloc[:, 1:].values != 0
+    n_expr_over_zero = pd.DataFrame(genes_not_equal_zero.sum(axis=0))
+    percent_expr_over_zero = (n_expr_over_zero / len(expression_df.index)) * 100
+    percent_expr_over_zero.index = expression_df.columns[1:]
+
+    # Filter genes with less than 10 percent individuals with non-zero expression
+    atleast10percent = percent_expr_over_zero[(percent_expr_over_zero > 10)[0]]
+    sample_ids = expression_df['sampleid']
+    expression_df = expression_df[atleast10percent.index]
+    expression_df.insert(loc=0, column='sampleid', value=sample_ids)
+    gene_ids = list(expression_df.columns.values)[1:]
+    geneloc_df = pd.read_csv(
+        f'gs://{INPUT_BUCKET}/kat/input/geneloc_chr22.tsv', sep='\t'
+    )
+    geneloc_df = geneloc_df[geneloc_df.geneid.isin(gene_ids)]
+
+    return len(geneloc_df.index)
+
+
+# Run Spearman rank in parallel by sending genes in a batches
+def run_computation_in_scatter(idx):  # pylint: disable=too-many-locals
+    """Run genes in scatter"""
+
+    expression_df = pd.read_csv(
```

Comment:
I don't yet have these files, but I just realised that this script will have to be adapted to handle different cell types and different chromosomes, likely all in the same folder but with different prefixes/chromosome numbers. E.g., `Plasma_chr22_log_residuals.tsv` might be `Bmem_chr21_log_residuals.tsv`. This is true for the largely all input files (e.g., expression_df, genotype_df, geneloc_df, snploc_df. Lastly, for the covariate_df, the cell type changes (e.q., `Plasma_peer_factors.tsv` to `Bmem_peer_factors.tsv`), but the chromosome number does not. @illusional would you have any ideas of how to do this?

Reply:
Hey @KatalinaBobowik! We can definitely make it do this, we hard coded a few things (because it was easier), but it's probably a good time to parameterise these inputs so they're more dynamic. If you're okay with it, I can make a branch off this branch, and open a pull request parameterising the inputs :)

Otherwise more than happy to jump on a zoom to work through it together.

Reply:
Making a branch off my branch sounds like a good idea. It'll be super helpful to see how to make the code more flexible/dynamic :)